### ISSUE TITLE
Deserialization based on type field

### ISSUE THREAD

**Issue by mikeholler** [marshmallow-code/marshmallow#1471]
I'm assessing using this library for our company's SDK and I am curious to know whether marshmallow supports deserialization based on type. I've searched the documentation and have not been able to find anything about this, but I know different tools use different terminology, so I might be missing something.

I'm coming from a Java background, where we have a similar, popular deserialization tool called Jackson with a feature called Polymorphic Deserialization. [[1]] Essentially, it works like this:

```kotlin
@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.PROPERTY,
    property = "type",
    visible = false
)
@JsonSubTypes(
    JsonSubTypes.Type(value = Cat::class, name = "Cat"),
    JsonSubTypes.Type(value = Dog::class, name = "Dog")
)
sealed class Pet {
    abstract val name: String
}

data class Dog(
    override val name: String,
    val isBarking: Boolean
): Pet

data class Cat(
    override val name: String,
    val isMeowing: Boolean
): Pet
```

The above code will allow deserialization like so:

```kotlin
// Returns an instance of Dog
ObjectMapper.readValue<Pet>("""
    {
        "type": "Dog",
        "name": "Fido",
        "isBarking": false
    }
""")

// Returns an instance of Cat
ObjectMapper.readValue<Pet>("""
    {
        "type": "Cat",
        "name": "Sparkle",
        "isMeowing": false
    }
""")

// Raises an exception -- mouse is not a valid pet
ObjectMapper.readValue<Pet>("""
    {
        "type": "Mouse",
        "name": "Squeak"
    }
""")

// Raises an exception -- dogs can't meow
ObjectMapper.readValue<Pet>("""
    {
        "type": "Dog",
        "name": "Tom",
        "isMeowing": true
    }
""")
```

Is there a way to do something similar with Marshmallow? I'm also interested in using `marshmallow_dataclass` as well (if that helps).

[1]: https://stackoverflow.com/questions/11798394/polymorphism-in-jackson-annotations-jsontypeinfo-usage

**Comment by lafrech** [comment]
[Polymorphism](https://github.com/marshmallow-code/marshmallow/issues?utf8=%E2%9C%93&q=label%3Apolymorphism+) is an appropriate term.

TL;DR

This is not in marshmallow core. There are (at least) two third-party libs out there providing this.

**Comment by mikeholler** [comment]
Thank you! I should have seen these -- don't know how I missed them. Especially with there being a label. I'll let those other questions stand and close this.
